datastream-common/src/main/java/com/linkedin/datastream/metrics/DynamicMetricsManager.java

```diff
@@ -129,60 +129,31 @@ private void updateCache(String simpleClassName, String key, String metricName,
    */
   @SuppressWarnings("unchecked")
   private <T extends Metric> T getMetric(String name, Class<T> clazz) {
+    Validate.notNull(clazz, "metric class argument is null.");
     if (clazz.equals(Counter.class)) {
       return (T) _metricRegistry.counter(name);
     } else if (clazz.equals(Meter.class)) {
       return (T) _metricRegistry.meter(name);
     } else if (clazz.equals(Histogram.class)) {
       return (T) _metricRegistry.histogram(name);
     } else if (clazz.equals(Gauge.class)) {
-      return (T) new ResettableGauge<>();
+      throw new IllegalArgumentException("This method doesn't work with Gauges.");
     } else if (clazz.equals(Timer.class)) {
-      return (T) new Timer();
+      return (T) _metricRegistry.timer(name);
     } else {
```

Comment:
no, but both Timer and Gauge were done this way (`new T()`) instead of actually registering with the underlying MetricRegistry. It seems that any attempt to register a Gauge or Timer wasn't really doing anything prior to this change.

Reply:
What's the reason gauge registrations are illegal here?